File: 49.group-anagrams/ga.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>

class Solution {
public:
    vector< vector<string> > groupAnagrams(vector<string>& strs) {
        unordered_map< string, multiset<string> > mp;
        for (string s : strs) {
            string t = s;
            sort(t.begin(), t.end());
            mp[t].insert(s);
        }

        vector< vector<string> > result;
        for (auto m : mp) {
            vector<string> anagram(m.second.begin(), m.second.end());
            result.push_back(anagram);
        }

        return result;
    }
};

File: 49.group-anagrams/ga.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    var mp = {};
    strs.forEach(function (ele) {
        var s = ele.split('').sort().join('');
        if (mp[s])
            mp[s].push(ele);
        else {
            mp[s] = [];
            mp[s].push(ele);
        }
    });

    var result = [];
    for (var key in mp) {
            result.push(mp[key]);
    }
    return result;
};

console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));